flink-runtime/src/main/java/org/apache/flink/runtime/state/StateBackendLoader.java

```diff
@@ -358,17 +360,51 @@ public static boolean stateBackendFromApplicationOrConfigOrDefaultUseManagedMemo
         return false;
     }
 
+    /**
+     * Load state backend which may wrap the original state backend for recovery.
+     *
+     * @param originalStateBackend StateBackend loaded from application or config.
+     * @param classLoader User code classloader.
+     * @param keyedStateHandles The state handles for restore.
+     * @return Wrapped state backend for recovery.
+     * @throws DynamicCodeLoadingException Thrown if keyed state handles of wrapped state backend
+     *     are found and the class was not found or could not be instantiated.
+     */
+    public static StateBackend loadStateBackendFromKeyedStateHandles(
+            StateBackend originalStateBackend,
+            ClassLoader classLoader,
+            Collection<KeyedStateHandle> keyedStateHandles)
+            throws DynamicCodeLoadingException {
+        if (keyedStateHandles.stream()
+                        .anyMatch(stateHandle -> stateHandle instanceof ChangelogStateBackendHandle)
+                && !isChangelogStateBackend(originalStateBackend)) {
+            return loadChangelogStateBackend(originalStateBackend, classLoader, true);
+        }
+        return originalStateBackend;
+    }
+
+    private static boolean isChangelogStateBackend(StateBackend backend)
+            throws DynamicCodeLoadingException {
+        try {
+            return Class.forName(CHANGELOG_STATE_BACKEND) == backend.getClass();
+        } catch (ClassNotFoundException e) {
+            throw new DynamicCodeLoadingException(
+                    "Cannot find DelegateStateBackend class: " + CHANGELOG_STATE_BACKEND, e);
+        }
+    }
+
     private static StateBackend loadChangelogStateBackend(
-            StateBackend backend, ClassLoader classLoader) throws DynamicCodeLoadingException {
+            StateBackend backend, ClassLoader classLoader, boolean forSwitch)
+            throws DynamicCodeLoadingException {
 
         // ChangelogStateBackend resides in a separate module, load it using reflection
         try {
             Constructor<? extends DelegatingStateBackend> constructor =
                     Class.forName(CHANGELOG_STATE_BACKEND, false, classLoader)
                             .asSubclass(DelegatingStateBackend.class)
-                            .getDeclaredConstructor(StateBackend.class);
+                            .getDeclaredConstructor(StateBackend.class, boolean.class);
             constructor.setAccessible(true);
-            return constructor.newInstance(backend);
+            return constructor.newInstance(backend, forSwitch);
```

Comment:
I think it's better to introduce a new class and instantiate it if `forSwitch == true`, so that:
1. The already fragile contract of constructor, accepting a nested arguement is not extended
2. The two classes have more clear responsibilities (read-only / normal restore) and there is no conditional logic inside `ChangelogStateBackend`; and thus, less complexity

Reply:
Yeah, I think your suggestion makes sense.
I just use 'ChangelogStateBackendForRecovery' to support recovery and switch.
But I haven't gotten a better name about it. Do you have any suggestions?

Reply:
The difficult part :) 
How about `DeactivatedChangelogStateBackend`?

Reply:
I have not found a better name. Just use this now.